pycoast/tests/test_pycoast.py

```diff
@@ -878,83 +879,127 @@ def test_foreground(self):
                             'width': 1.0, 'minor_width': 0.5, 'minor_is_tick': True,
                             'write_text': True, 'lat_placement': 'lr', 'lon_placement': 'b'}
-        img = cw.add_overlay_from_dict(overlays, area_def)
-        os.remove(os.path.join(tmp, 'pycoast_cache_fakearea.png'))
+        cw.add_overlay_from_dict(overlays, area_def)
+        os.remove(cache_filename)
+
+    def test_caching_with_param_changes(self, tmpdir):
+        """Testing caching when changing parameters."""
+        from pycoast import ContourWriterPIL
+
+        # img = Image.new('RGB', (640, 480))
+        proj4_string = \
+            '+proj=stere +lon_0=8.00 +lat_0=50.00 +lat_ts=50.00 +ellps=WGS84'
+        area_extent = (-3363403.31, -2291879.85, 2630596.69, 2203620.1)
+        area_def = FakeAreaDef(proj4_string, area_extent, 640, 480)
+        cw = ContourWriterPIL(gshhs_root_dir)
+
+        font = ImageFont.truetype(os.path.join(
+            os.path.dirname(__file__), 'test_data', 'DejaVuSerif.ttf'))
+        overlays = {'cache': {'file': os.path.join(tmpdir, 'pycoast_cache')},
+                    'grid': {'font': font}}
+
+        # Create the original cache file
+        cw.add_overlay_from_dict(overlays, area_def)
+        cache_glob = glob(os.path.join(tmpdir, 'pycoast_cache_*.png'))
+        assert len(cache_glob) == 1
+        cache_filename = cache_glob[0]
+        assert os.path.isfile(cache_filename)
+        mtime = os.path.getmtime(cache_filename)
+
+        # Reuse the generated cache file
+        cw.add_overlay_from_dict(overlays, area_def)
+        cache_glob = glob(os.path.join(tmpdir, 'pycoast_cache_*.png'))
+        assert len(cache_glob) == 1
+        assert os.path.isfile(cache_filename)
+        assert os.path.getmtime(cache_filename) == mtime
+
+        # Remove the font option, should produce the same result
+        # font is not considered when caching
+        del overlays['grid']['font']
+        cw.add_overlay_from_dict(overlays, area_def)
+        cache_glob = glob(os.path.join(tmpdir, 'pycoast_cache_*.png'))
+        assert len(cache_glob) == 1
+        assert os.path.isfile(cache_filename)
+        assert os.path.getmtime(cache_filename) == mtime
+
+        # Changing a parameter should create a new cache file
+        overlays = {'cache': {'file': os.path.join(tmpdir, 'pycoast_cache')},
+                    'grid': {'width': 2.0}}
+        cw.add_overlay_from_dict(overlays, area_def)
+        cache_glob = glob(os.path.join(tmpdir, 'pycoast_cache_*.png'))
+        assert len(cache_glob) == 2
+        assert os.path.isfile(cache_filename)
+        new_cache_filename = cache_glob[0] if cache_glob[0] != cache_filename else cache_glob[1]
+        # original cache file should be unchanged
+        assert os.path.getmtime(cache_filename) == mtime
+        # new cache file should be...new
+        assert os.path.getmtime(new_cache_filename) != mtime
```

Comment:
This should be split into individual test methods, probably with a common setup for the common initialisation.

Reply:
I knew you were going to say that. I copied this from your tests :wink: 

However, because this is caching these tests actually all depend on each other. The first run fills the cache, the next two cases check that calling again uses the same cache file, and the last one makes sure that when we change the parameters a new cache file is created. I *could* update the last one to only change the `width` parameter, but otherwise the rest of the stuff is needed in one method.

Reply:
ok

Reply:
> I knew you were going to say that. I copied this from your tests :wink:

Haha, that is not necessarily a mark of quality :D